Objects/unionobject.c

```diff
@@ -0,0 +1,433 @@
+    if (strcmp(type->tp_name, name) != 0) {
+        return 0;
+    }
+    PyObject *module = PyObject_GetAttrString((PyObject *)type, "__module__");
+    if (module == NULL) {
+        return -1;
+    }
+    int res = PyUnicode_Check(module)
+        && _PyUnicode_EqualToASCIIString(module, "typing");
+    Py_DECREF(module);
+    return res;
+}
+
+static PyMethodDef union_methods[] = {
+    {"__instancecheck__", union_instancecheck, METH_O},
+    {"__subclasscheck__", union_subclasscheck, METH_O},
+    {0}};
+
+static PyObject *
+union_richcompare(PyObject *a, PyObject *b, int op)
+{
+    if (op != Py_EQ && op != Py_NE) {
+        PyObject *result = Py_NotImplemented;
+        Py_INCREF(result);
+        return result;
+    }
+
+    PyObject* b_set = NULL;
+    unionobject *aa = (unionobject *)a;
+    PyObject* a_set = PySet_New(aa->args);
+
+    if (is_typing_name(b, "_UnionGenericAlias")) {
+        PyObject* b_args = PyObject_GetAttrString(b, "__args__");
+        b_set = PySet_New(b_args);
+    }
+
+    PyTypeObject *type = Py_TYPE(b);
+    if (type == &Py_UnionType) {
+        unionobject *bb = (unionobject *)b;
+        b_set = PySet_New(bb->args);
+    }
+
+    if (b_set == NULL) {
+        Py_RETURN_FALSE;
+    }
+
+    return PyObject_RichCompare(a_set, b_set, op);
+}
+
+static PyObject*
+flatten_args(PyObject* args)
+{
+    int arg_length = PyTuple_GET_SIZE(args);
+    int total_args = 0;
+    // Get number of total args once it's flattened.
+    for (int i = 0; i < arg_length; i++) {
+        PyObject *arg = PyTuple_GET_ITEM(args, i);
+        PyTypeObject* arg_type = Py_TYPE(arg);
+        if (arg_type == &Py_UnionType) {
+            unionobject *alias = (unionobject *)arg;
+            total_args = total_args + PyTuple_GET_SIZE(alias->args);
+        } else {
+            total_args++;
+        }
+    }
+    // Create new tuple of flattened args.
+    PyObject *flattened_args = PyTuple_New(total_args);
+    int pos = 0;
+    for (int i = 0; i < arg_length; i++) {
+        PyObject *arg = PyTuple_GET_ITEM(args, i);
+        PyTypeObject* arg_type = Py_TYPE(arg);
+        if (arg_type == &Py_UnionType) {
+            unionobject *alias = (unionobject *)arg;
+            PyObject* nested_args = alias->args;
+            int nested_arg_length = PyTuple_GET_SIZE(nested_args);
+            for (int j = 0; j < nested_arg_length; j++) {
+                PyObject* nested_arg = PyTuple_GET_ITEM(nested_args, j);
+                Py_INCREF(nested_arg);
+                PyTuple_SET_ITEM(flattened_args, pos, nested_arg);
+                pos++;
+            }
+        } else {
+            Py_INCREF(arg);
+            PyTuple_SET_ITEM(flattened_args, pos, arg);
+            pos++;
+        }
+    }
+    Py_INCREF(flattened_args);
+    return flattened_args;
+}
+
+static PyObject*
+dedup_and_flatten_args(PyObject* args)
+{
+    args = flatten_args(args);
+    int arg_length = PyTuple_GET_SIZE(args);
+    PyObject* temp[arg_length];
```

Comment:
This is causing issues in the windows build.

Is there a recommended workaround for variable length arrays? 

Alternatively, is there a maximum number of arguments one could have in a Union?

Reply:
These `args` will be subclasses of `type` no? Why not using a set to de-duplicate them (as subclasses of `type` will support `__hash__` and `__eq__`)?  Otherwise, this code will be `O(n**2)` on the number of types added.

Reply:
I'd prefer not to use a set here, because we ought to return non-duplicate items in the order they came in. It matters because we want `repr(int | str)` to return `int | str`, not occasionally `str | int`.

I'm not worried about the _O(n**2)_ behavior here, these loops are just doing pointer comparisons and typical unions don't have that many members. But I don't want to declare a fixed maximum size. So I recommend `PyMem_Calloc()` and `PyMem_Free()`. (Alternatively, you could allocate a tuple of size `arg_length`, store the pointers there, and at the end use `_PyTuple_Resize()`. But it wouldn't look cleaner.)